sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -374,7 +374,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       if (scale < _scale) {
         // Easier case: we just need to divide our scale down
```

Comment:
The code is hard to read now. Can we add a new branch for `diff > MAX_LONG_DIGITS`?
```
if (diff > MAX_LONG_DIGITS) {
  lv = 0
  roundMode match {
    case ROUND_FLOOR | ... => lv = -1
    case ROUND_CEILING | ... => lv = 1
  }
} else {
  original code
}
```

Reply:
Yeah, my initial version was similar to that. But then I thought that we are kind of duplicating the same logic in two branches and I decided to merge them. I'm fine with both versions though. Does the following look good?

```scala
        if (diff > MAX_LONG_DIGITS) {
          lv = roundMode match {
            case ROUND_FLOOR => if (lv < 0) -1L else 0L
            case ROUND_CEILING => if (lv > 0) 1L else 0L
            case ROUND_HALF_UP | ROUND_HALF_EVEN => 0L
            case _ => throw QueryExecutionErrors.unsupportedRoundingMode(roundMode)
          }
        } else {
          original code
        }
```

Reply:
This looks better to me.

Reply:
Let's also add code comments to explain why we need special handling for `diff > MAX_LONG_DIGITS`